Make the game (visually) run at framerates above ~30 FPS

InfoTeddy opened the pull request:
## Changes:

VVVVVV is a game that runs at 1000/34 frames per second, which is approximately 29.4 frames a second, which is close to, but not quite, 30 frames a second. This odd number comes from the fact that the amount of times it takes for each frame is coded to be 34 milliseconds, and not 33.333333333... milliseconds.

This pull request unlocks VVVVVV's framecap of 1000/34 FPS. It does this in such a way that the underlying engine still runs at an actual framerate of 1000/34 FPS (which we should be calling tickrate now), and in between each actual frame we just draw what the frames in between each frame *should* be, using something called "linear interpolation". So now the game will visually run at the same rate as your monitor (my monitor is 60hz, I'm not a true gamer :( &#x200b; &#x200b; &#x200b; &#x200b; ).

What's linear interpolation, you might ask? Well it's basically like drawing a straight line between two points on a graph. But it's more akin to averaging something. All the rendering does is go like, "Ok so on this frame, we have 2, and on the next frame, we have 4. What's the average of 2 and 4? Why, it's 3! So we should draw a frame using 3."

This approach means that the underlying physics isn't changed, which is a very good thing because if I changed the physics around to update at, say, twice the speed but with half the values, that would run the risk of breaking custom levels. Also, it's easier to do it this way because it requires changing less values, although it does require disentangling all pieces of update code from render code so as to make sure things don't update as quickly as possible. And regardless of those concerns, this approach is the cleanest imo, since the physics don't subtly change depending on your refresh rate (unlike games like Super Meat Boy, which apparently [desyncs TASes depending on if your refresh rate is 59hz or 60hz](/clementgallet/libTAS/issues/220), oh and also different resolutions have different refresh rates, good luck).

Oh, and by the way, the Switch version (made by some weird company named Nicalis) actually does the screw-up-the-physics approach instead of what I'm doing. They've apparently had to change some custom levels because the way they did it broke things like edge-flipping. Maddening.

Most of the work I took was from [Colon-D's fork of VVVVVV attempting to run it at over 30 FPS](/Colon-D/VVVVVV-over-30-FPS). However, his mod is incomplete (for example, the title screen still runs at 1000/34 FPS) and also does some things inefficiently (like re-drawing warp backgrounds every single frame, instead of scrolling a surface + incoming textures after drawing an initial background). Also it hasn't been updated in months.

This patch also seems to improve the 100 milliseconds of `SDL_Delay` the unfocus-pause gives, in that it actually seems to have an effect now and be noticeable when you re-focus (possibly because before we were using `SDL_Delay` as the frame limiter, too, and I guess one of them kind of canceled the other out?).

To enable this, you need to go into "graphic options" and turn on the "toggle fps" option. You can turn on "toggle vsync" as well if you feel like it, too.

Honestly, after playing the game at 60 FPS, it feels really strange to play the game at normal 30 FPS.

#99 is already closed, but seems relevant to mention in this PR.

(Current base commit ID for this PR: 62441edbc9b279d4e37ea6a640f376288ffc8417)

## Legal Stuff:

By submitting this pull request, I confirm that...

- [X] My changes may be used in a future commercial release of VVVVVV (for
  example, a 2.3 update on Steam for Windows/macOS/Linux)
- [X] I will be credited in a `CONTRIBUTORS` file and the "GitHub Friends"
  section of the credits for all of said releases, but will NOT be compensated
  for these changes

flibitijibibo commented:
Wow, that was fast!

It'll take me a while to get through the technical stuff, but @TerryCavanagh will be the expert on accuracy here, so I'll defer to him on how it plays and what doesn't feel right. I'm going to guess and say we'll want this off by default for preservation's sake, but it'll probably resolve the biggest complaint the game's ever had if we at least have the option...

InfoTeddy commented:
> I'm going to guess and say we'll want this off by default for preservation's sake

Yeah, I figured we should have a framerate-limiting option. I tried re-adding the old frame limiting system in, but it felt a bit dirty to be doing that, what with the `SDL_Delay` and all. I'm thinking the option should just be a toggle between "30" FPS and v-sync, no?

flibitijibibo commented:
Makes sense to me! I’ll start going through the commits on Monday.

flibitijibibo commented:
Ran through all the commits and they seem pretty solid - some of them could probably be split off into their own pull request but it's not really important.

At this point, we need people to test this _really_ thoroughly - if you're reading this and know how to compile the game, that means you. As long as it looks and plays sanely, both in 60 and 30 mode, this should be good to go.

InfoTeddy commented:
I fixed the preloader. I wanted to interpolate it until I found out that there's no movement and it's actually somewhat random each time, so. Also I rebased onto the latest upstream and force-pushed.

InfoTeddy commented:
@flibitijibibo Would applying #22 (after fixing conflicts) be a good method of v-sync? Also I'm thinking now that v-sync should be its own separate option, and 30/above-30 another one. If so then I can't be using sleep() -> update() -> sleep() to limit the framerate to 30.

flibitijibibo commented:
If it's possible to have both that'd work for me. The way I _like_ to have it in games (when possible) is to have both a fixed-step and vsync option, since fixed-step is good for determinism while vsync is good for visuals, and being able to toggle both cases is good for prioritizing accuracy vs. visuals without getting stuck with stuttering or weird lerpy glitches, for machines where their display isn't _exactly_ the same as the timestep. Dust: AET is usually my go-to example for this. For this game, having 30 or 30+ as well as sync/tear makes sense, since there are reasonable use cases for all 4 combinations:

30+Sync: Original visuals on a compatible monitor, no tearing
30+Tear: Original visuals and avoids stuttering
Unlocked+Sync: The ideal scenario
Unlocked+Tear: Better than 30+Tear visually and avoids stuttering, just maybe don't do this on a hot day...

ThatWeirdAndrew commented:
I've played both version for about 30 minutes now and have not noticed any issues so far.

AllyTally commented:
I have mentioned this on the VVVVVV Discord server, but it seems possible to get teleported to (20, 20) by going through a room that wraps. [Here's a video](https://o.lol-sa.me/2mMWFmd.mp4), and [this is possible with it](https://o.lol-sa.me/4onmV45.png). It's a pretty game breaking bug, as I have been teleported to (20, 20) multiple times while trying to play through the Warp Zone normally.

I'm making this a comment here because from what I can tell, this only happens for this pull request. I also compiled it on Windows 10 (which is how I found #224)

InfoTeddy commented:
I have also mentioned this on the VVVVVV Discord server, but this sounds like plain old Undefined Behavior that may just manifest from shifting memory around the right way.

parkerlreed commented:
Looking good! 500FPS VVVVVV is not something I ever expected to see 

![image](https://user-images.githubusercontent.com/841440/81095672-5dbafa80-8ed3-11ea-8836-851f2b21e830.png)

parkerlreed commented:
Pointed out on Discord but should put it here for posterity. 

Intel GPU on Linux with Mesa doesn't Vsync to monitor 
![image](https://user-images.githubusercontent.com/841440/81097455-f0f52f80-8ed5-11ea-93e1-60e40cfed4aa.png)

flibitijibibo commented:
The vsync option can be labeled as an SDL issue, the problem is that SDL_Render does not listen to the hint as it should:

https://hg.libsdl.org/SDL/file/fdabcc76aaaf/src/render/SDL_render.c#l789
https://hg.libsdl.org/SDL/file/fdabcc76aaaf/src/render/opengl/SDL_render_gl.c#l1648

A workaround would be recreating the renderer, but that's a sloppy fix for something that SDL can easily listen for with an SDL_HintCallback and send to the backends, all of which trivially support dynamic swap intervals. I haven't checked if this is filed on the Bugzilla, but if not we should get that done on SDL's end.

leo60228 commented:
I'm not sure if "as it should" is the right thing to say here. SDL_HINT_RENDER_VSYNC is only used by SDL_CreateRenderer, and the documentation doesn't really suggest otherwise.

InfoTeddy commented:
(Just rebased onto latest master and force-pushed, so #224 now applies to this branch.)

leo60228 commented:
I'm trying to fix this in SDL, but I haven't used Mercurial before, so that's making it somewhat annoying. For now, maybe just change `main` to:
```c++
        game.init();
        graphics.init();
        game.loadstats();
        Screen gameScreen;
        graphics.screenbuffer = &gameScreen;
```
and add a note saying that a restart is required. Though, `game.loadstats()` mutates `gameScreen`... Maybe add null checks and call it twice?

InfoTeddy commented:
(Just rebased onto latest master and force-pushed, so #227 now applies to this branch.)

InfoTeddy commented:
(Just rebased and force-pushed, this is to fix the `<vsync>` option in `unlock.vvv` not being loaded properly.)

leo60228 commented:
https://bugzilla.libsdl.org/show_bug.cgi?id=5123

InfoTeddy commented:
(Just rebased onto latest master, so #228 now applies to this PR.)

InfoTeddy commented:
(Rebased onto latest master, so #229 now applies to this patch.)

InfoTeddy commented:
I've rebased onto latest master, so #230 and #231 now apply to this PR. Additionally I've also made sure no interpolation occurs in 30-FPS-only mode, and fixed an off-by-one in the fade-to-menu/lab delays.

InfoTeddy commented:
Rebased and force-pushed, #232, #234, and #235 now apply to this branch. (Also a re-fix of #235 due to game loop changes.)

InfoTeddy commented:
(Rebased, #238 now applies.)

InfoTeddy commented:
(Rebased, #239, #240, and #241 now apply.)

InfoTeddy commented:
(Rebased, #242 applies.)

InfoTeddy commented:
(Rebased, #243, #244, and #245 now apply.)

InfoTeddy commented:
(Rebased, #246 and #247 now applies.)

InfoTeddy commented:
(Rebased, #248, #249, #250, #251, and #252 apply.)

InfoTeddy commented:
(Rebased, #143 now applies.)

InfoTeddy commented:
(Rebased, #253, #254, and #255 apply.)

InfoTeddy commented:
(Rebased, #243 applies.)

leo60228 commented:
Rebasing every time master changes creates a bunch of pointless notifications. Maybe you could rebase locally and only push when something actually needs it? Or push the rebases to a different branch and `git reset --hard other-branch && git push --force`?

flibitijibibo commented:
I haven't been getting notifications other than posts saying that the branch has been rebased... is there a specific setting that notifies per-commit? I do get emails for unique commits, but force-pushes don't show up for me.

leo60228 commented:
Those are the ones I'm getting. Now that you mention it, maybe they should just be edited into the PR itself?

InfoTeddy commented:
Alright, I'll get rid of all the rebase posts and put a base commit ID in the description. Every time I rebase, I'll edit the description to update that commit ID.

Fussmatte commented:
I've went through the first few levels of Dimension Open with this version and it works awesome! I don't notice any issue so far.

flibitijibibo commented:
Making a last call for testing this! I'll be running through the game this week, and if I find nothing wrong I'll merge this on the 19th.

flibitijibibo commented:
Did a 100% run on a single program instance and everything looked fine to me! I did notice that when running with Valgrind, performance was worse no matter which mode I was running with. On an i7 2600K I can maintain 30fps with Valgrind using the master branch, but couldn't get the same with this branch. We could probably get those frames back by optimizing BlitSurfaceColoured (it's a very particular type of blend operation though, blech), but it'd be worth checking this branch out on an obscenely bad PC first to make sure we're not accidentally bumping the system requirements.

InfoTeddy commented:
Mm, that could be down to no longer using `SDL_Delay()` to limit the framerate. I'm pretty sure that in 30-FPS mode, it does the rendering at 30 FPS too (it doesn't like, render unnecessary delta frames), so I'm not sure where the performance drop comes from.

I tested it out in Valgrind myself. On master, it's very stuttery 30 FPS, and on this branch it seems to be 15 FPS or so, in both modes. (I had to make `musicclass::play()` and `musicclass::playef()` do nothing though, otherwise Valgrind complains about an illegal opcode in Vorbis stuff in SDL2_mixer.) Although on both branches, there's a very large startup time (but there's also a very large startup time whenever I run the game under GDB, too, so).

flibitijibibo commented:
>Mm, that could be down to no longer using SDL_Delay() to limit the framerate.

Thinking about it, that's probably why the music thread is falling behind too; the sleep probably gives the VM the room it needs to mix everything in time. As long as a crappy PC doesn't die with the changes that's good enough for me!

InfoTeddy commented:
I have a terrible enough computer that zip decompression hurts to do...
![Zip decompression hurts my computer](https://user-images.githubusercontent.com/59748578/84695398-a072e800-aeff-11ea-94a9-dcb2d683d305.png)
...but this patch runs fine on my system. So I think it's fine.

flibitijibibo commented:
Beautiful. So as long as nobody finds anything else this week we're good to go for Friday's merge.

parkerlreed commented:
What option enables this for master?

InfoTeddy commented:
Go to Graphic Options, there will be an option labeled Toggle FPS.

parkerlreed commented:
Oh I see over 30 fps in in the graphics options. Sweet
